File: scripts/dcgan.py
import torch
import torch.nn as nn
import torch.optim as optim
import numpy as np
import torchvision
from torchvision import transforms, datasets, utils
import os
import matplotlib.pyplot as plt

# Size of z latent vector (i.e. size of generator input)
latent_input_size = 100

# Size of feature maps in generator
image_size = 28

# Number of training epochs
num_epochs = 5

# Learning rate for optimizers
lr = 0.0002

# Beta1 hyperparam for Adam optimizers
beta1 = 0.5


transform = transforms.Compose([
                               transforms.ToTensor(),
                           ])

kmnist = torchvision.datasets.KMNIST("data/kmnist", train=True, transform=transform, download=True)


dataloader = torch.utils.data.DataLoader(kmnist, batch_size=128,
                                         shuffle=True, num_workers=4)

# Decide which device we want to run on
device = torch.device("cuda:0" if torch.cuda.is_available()  else "cpu")

# Plot some training images
real_batch = next(iter(dataloader))
real_batch[0].shape
plt.figure(figsize=(8,8))
plt.axis("off")
plt.title("Training Images")
plt.imshow(np.transpose(utils.make_grid(real_batch[0].to(device)[:64], padding=2, normalize=True).cpu(),(1,2,0)))
plt.show()

def weights_init(m):
    classname = m.__class__.__name__
    if classname.find('Conv') != -1:
        nn.init.normal_(m.weight.data, 0.0, 0.02)
    elif classname.find('BatchNorm') != -1:
        nn.init.normal_(m.weight.data, 1.0, 0.02)
        nn.init.constant_(m.bias.data, 0)
        
# Generator Code

class Generator(nn.Module):
    def __init__(self):
        super(Generator, self).__init__()
        self.main = nn.Sequential(
            # torch.nn.ConvTranspose2d(in_channels, out_channels, kernel_size, stride=1, padding=0, output_padding=0, groups=1, bias=True, dilation=1, padding_mode='zeros')
            # h_out = (h_in - 1) * stride - 2 * padding + dilation * (kernel_size - 1) + output_padding + 1
            # (1 - 1) * 1 - 2 * 0 + 1 * (4 -1 ) + 0 + 1
            # 4 x 4
            nn.ConvTranspose2d(latent_input_size, image_size * 4, 4, 1, 0, bias=False),
            nn.BatchNorm2d(image_size * 4),
            nn.ReLU(True),
            # 8 * 8
            nn.ConvTranspose2d(image_size * 4, image_size * 2, 4, 2, 1, bias=False),
            nn.BatchNorm2d(image_size * 2),
            nn.ReLU(True),
            # 16 x 16
            nn.ConvTranspose2d(image_size * 2, image_size, 4, 2, 2, bias=False),
            nn.BatchNorm2d(image_size),
            nn.ReLU(True),
            # 28 x 28
            nn.ConvTranspose2d(image_size, 1, 4, 2, 1, bias=False),
            nn.Tanh()
        )
    def forward(self, input):
        return self.main(input)

generator = Generator().to(device)

# Apply the weights_init function to randomly initialize all weights
#  to mean=0, stdev=0.2.
generator.apply(weights_init)

class Discriminator(nn.Module):
    def __init__(self):
        super(Discriminator, self).__init__()
        self.main = nn.Sequential(
            # 14 x 14
            nn.Conv2d(1, image_size, 4, 2, 1, bias=False),
            nn.LeakyReLU(0.2, inplace=True),
            # 7 x 7
            nn.Conv2d(image_size, image_size * 2, 4, 2, 1, bias=False),
            nn.BatchNorm2d(image_size * 2),
            nn.LeakyReLU(0.2, inplace=True),
            # 3 x 3
            nn.Conv2d(image_size * 2, image_size * 4, 4, 2, 1, bias=False),
            nn.BatchNorm2d(image_size * 4),
            nn.LeakyReLU(0.2, inplace=True),
            # 1 x 1
            nn.Conv2d(image_size * 4, 1, 4, 2, 1, bias=False),
            nn.Sigmoid()
        )
    def forward(self, input):
        return self.main(input)


discriminator = Discriminator().to(device)

# Apply the weights_init function to randomly initialize all weights
#  to mean=0, stdev=0.2.
discriminator.apply(weights_init)


# Initialize BCELoss function
criterion = nn.BCELoss()

# Create batch of latent vectors that we will use to visualize
#  the progression of the generator
fixed_noise = torch.randn(64, latent_input_size, 1, 1, device=device)

# Establish convention for real and fake labels during training
real_label = 1
fake_label = 0

# Setup Adam optimizers for both G and D
disc_optimizer = optim.Adam(discriminator.parameters(), lr=lr, betas=(beta1, 0.999))
gen_optimizer = optim.Adam(generator.parameters(), lr=lr, betas=(beta1, 0.999))

# Training Loop

# Lists to keep track of progress
img_list = []
generator_losses = []
discriminator_losses = []

print("Starting Training Loop...")
# For each epoch
for epoch in range(num_epochs):
    # For each batch in the dataloader
    for i, data in enumerate(dataloader, 0):
        ############################
        # (1) Update D network: maximize log(D(x)) + log(1 - D(G(z)))
        ###########################
        ## Train with all-real batch
        discriminator.zero_grad()
        # Format batch
        real_cpu = data[0].to(device)
        b_size = real_cpu.size(0)
        label = torch.full((b_size,), real_label, device=device)
        # Forward pass real batch through D
        output = discriminator(real_cpu).view(-1)
        # Calculate loss on all-real batch
        discriminator_loss_real = criterion(output, label)
        # Calculate gradients for D in backward pass
        discriminator_loss_real.backward()
        discriminator_real_verdict = output.mean().item()
        ## Train with all-fake batch
        # Generate batch of latent vectors
        noise = torch.randn(b_size, latent_input_size, 1, 1, device=device)
        # Generate fake image batch with G
        fake = generator(noise)
        label.fill_(fake_label)
        # Classify all fake batch with D
        output = discriminator(fake.detach()).view(-1)
        # Calculate D's loss on the all-fake batch
        discriminator_loss_fake = criterion(output, label)
        # Calculate the gradients for this batch
        discriminator_loss_fake.backward()
        # Add the gradients from the all-real and all-fake batches
        discriminator_loss = discriminator_loss_real + discriminator_loss_fake
        # Update D
        disc_optimizer.step()
        ############################
        # (2) Update G network: maximize log(D(G(z)))
        ###########################
        generator.zero_grad()
        label.fill_(real_label)  # fake labels are real for generator cost
        # Since we just updated D, perform another forward pass of all-fake batch through D
        output = discriminator(fake).view(-1)
        # Calculate G's loss based on this output
        generator_loss = criterion(output, label)
        # Calculate gradients for G
        generator_loss.backward()
        discriminator_fake_verdict = output.mean().item()
        # Update G
        gen_optimizer.step()
        # Output training stats
        if i % 50 == 0:
            print('[%d/%d][%d/%d]\tLoss_D: %.4f\tLoss_G: %.4f\tD(x): %.4f\tD(G(z)): %.4f '
                  % (epoch, num_epochs, i, len(dataloader),
                     discriminator_loss.item(), generator_loss.item(), discriminator_real_verdict, discriminator_fake_verdict))
        # Save Losses for plotting later
        generator_losses.append(generator_loss.item())
        discriminator_losses.append(discriminator_loss.item())
        # Check how the generator is doing by saving G's output on fixed_noise
        with torch.no_grad():
          fake = generator(fixed_noise).detach().cpu()
          img_list.append(utils.make_grid(fake, padding=2, normalize=True))

plt.figure(figsize=(10,5))
plt.title("Generator and Discriminator Loss During Training")
plt.plot(generator_losses,label="G")
plt.plot(discriminator_losses,label="D")
plt.xlabel("iterations")
plt.ylabel("Loss")
plt.legend()
plt.show()

fig = plt.figure(figsize=(8,8))
plt.axis("off")
ims = [[plt.imshow(np.transpose(i,(1,2,0)))] for i in img_list]
plt.show()

# Grab a batch of real images from the dataloader
real_batch = next(iter(dataloader))

# Plot the real images
plt.figure(figsize=(15,15))
plt.subplot(1,2,1)
plt.axis("off")
plt.title("Real Images")
plt.imshow(np.transpose(utils.make_grid(real_batch[0].to(device)[:64], padding=5, normalize=True).cpu(),(1,2,0)))
plt.show()
# Plot the fake images from the last epoch
plt.subplot(1,2,2)
plt.axis("off")
plt.title("Fake Images")
plt.imshow(np.transpose(img_list[-1],(1,2,0)))
plt.show()

File: scripts/dcgan.R
library(torch)


# Data loading ------------------------------------------------------------


dir <- "/tmp"

batch_size <- 128

kmnist <- kmnist_dataset(
    dir,
    download = TRUE,
    transform = function(x) {
        x <- x$to(dtype = torch_float())/256
        x <- 2*(x - 0.5)
        x[newaxis,..]
    }
)
dl <- dataloader(kmnist, batch_size = batch_size, shuffle = TRUE)


# Model -------------------------------------------------------------------


device <- if (cuda_is_available()) torch_device("cuda:0") else "cpu"

latent_input_size <- 100
image_size <- 28

generator <- nn_module(
    "generator",
    initialize = function() {
        self$main = nn_sequential(
            # nn_conv_transpose2d(in_channels, out_channels, kernel_size, stride=1, padding=0, output_padding=0, groups=1, bias=TRUE, dilation=1, padding_mode='zeros')
            # h_out = (h_in - 1) * stride - 2 * padding + dilation * (kernel_size - 1) + output_padding + 1
            # (1 - 1) * 1 - 2 * 0 + 1 * (4 -1 ) + 0 + 1
            # 4 x 4
            nn_conv_transpose2d(latent_input_size, image_size * 4, 4, 1, 0, bias = FALSE),
            nn_batch_norm2d(image_size * 4),
            nn_relu(),
            # 8 * 8
            nn_conv_transpose2d(image_size * 4, image_size * 2, 4, 2, 1, bias = FALSE),
            nn_batch_norm2d(image_size * 2),
            nn_relu(),
            # 16 x 16
            nn_conv_transpose2d(image_size * 2, image_size, 4, 2, 2, bias = FALSE),
            nn_batch_norm2d(image_size),
            nn_relu(),
            # 28 x 28
            nn_conv_transpose2d(image_size, 1, 4, 2, 1, bias = FALSE),
            nn_tanh()
        )
    },
    forward = function(x) {
        self$main(x)
    }
)

gen <- generator()

discriminator <- nn_module(
    "discriminator",
    initialize = function() {
        self$main = nn_sequential(
            # 14 x 14
            nn_conv2d(1, image_size, 4, 2, 1, bias = FALSE),
            nn_leaky_relu(0.2, inplace = TRUE),
            # 7 x 7
            nn_conv2d(image_size, image_size * 2, 4, 2, 1,  bias = FALSE),
            nn_batch_norm2d(image_size * 2),
            nn_leaky_relu(0.2, inplace = TRUE),
            # 3 x 3
            nn_conv2d(image_size * 2, image_size * 4, 4, 2, 1,  bias = FALSE),
            nn_batch_norm2d(image_size * 4),
            nn_leaky_relu(0.2, inplace = TRUE),
            # 1 x 1
            nn_conv2d(image_size * 4, 1, 4, 2, 1,  bias = FALSE),
            nn_sigmoid()
        )
    },
    forward = function(x) {
        self$main(x)
    }
)

disc <- discriminator()

init_weights <- function(m) {
    if (grepl("conv", m$.classes[[1]])) {
        nn_init_normal_(m$weight$data(), 0.0, 0.02)
    } else if (grepl("batch_norm", m$.classes[[1]])) {
        nn_init_normal_(m$weight$data(), 1.0, 0.02)
        nn_init_constant_(m$bias$data(), 0)
    }
}

gen[[1]]$apply(init_weights)

disc[[1]]$apply(init_weights)

gen$to(device = device)
disc$to(device = device)


# Training ----------------------------------------------------------------


criterion <- nn_bce_loss()

learning_rate <- 0.0002

disc_optimizer <- optim_adam(disc$parameters, lr = learning_rate, betas = c(0.5, 0.999))
gen_optimizer <- optim_adam(gen$parameters, lr = learning_rate, betas = c(0.5, 0.999))

fixed_noise <- torch_randn(c(64, latent_input_size, 1, 1), device = device)
num_epochs <- 5

img_list <- vector(mode = "list", length = num_epochs * trunc(dl$.iter()$.length()/50))
gen_losses <- c()
disc_losses <- c()

normalize <- function(x) {
    min = x$min()$item()
    max = x$max()$item()
    x$clamp_(min = min, max = max)
    x$add_(-min)$div_(max - min + 1e-5)
    x
}

# 4D mini-batch Tensor of shape (B x C x H x W)
make_grid <- function(tensor, num_rows = 8, padding = 2, pad_value = 0) {
    nmaps <- tensor$size(0)
    xmaps <- min(num_rows, nmaps)
    ymaps <- ceiling(nmaps/xmaps)
    height <- floor(tensor$size(2) + padding)
    width <- floor(tensor$size(3) + padding)
    num_channels <- tensor$size(1)
    grid <- tensor$new_full(c(num_channels, height * ymaps + padding, width * xmaps + padding), pad_value)
    k <- 0
    for (y in 0:(ymaps - 1)) {
        for (x in 0:(xmaps - 1)) {
            if (k >= nmaps) break
            grid$narrow(
                dim = 1,
                start = torch_tensor(y * height + padding, dtype = torch_int64())$sum(dim = 0),
                length = height - padding)$narrow(
                    dim = 2,
                    start = torch_tensor(x * width + padding, dtype = torch_int64())$sum(dim = 0),
                    length = width - padding)$copy_(tensor[k + 1, , , ])
            k <- k + 1
        }
    }
    grid
}

img_num <- 0
for (epoch in 1:num_epochs) {

    batchnum <- 0
    for (b in enumerate(dl)) {

        batchnum <- batchnum + 1

        y_real <- torch_ones(b[[1]]$size()[1], device = device)
        y_fake <- torch_zeros(b[[1]]$size()[1], device = device)

        noise <- torch_randn(b[[1]]$size()[1], latent_input_size, 1, 1, device = device)
        fake <- gen(noise)
        img <- b[[1]]$to(device = device)

        # update discriminator
        disc_loss <- criterion(disc(img), y_real) + criterion(disc(fake$detach()), y_fake)

        disc_optimizer$zero_grad()
        disc_loss$backward()
        disc_optimizer$step()

        # update generator
        gen_loss <- criterion(disc(fake), y_real)

        gen_optimizer$zero_grad()
        gen_loss$backward()
        gen_optimizer$step()

        disc_losses <- c(disc_losses, disc_loss$cpu()$item())
        gen_losses <- c(gen_losses, gen_loss$cpu()$item())

        if (batchnum %% 50 == 0) {
            img_num <- img_num + 1
            cat("Epoch: ", epoch,
                "    batch: ", batchnum,
                "    disc loss: ", as.numeric(disc_loss$cpu()),
                "    gen loss: ", as.numeric(gen_loss$cpu()),
                "\n")
            with_no_grad({
                generated <- gen(fixed_noise)
                grid <- make_grid(normalize(generated))
                img_list[[img_num]] <- as_array(grid$to(device = "cpu"))
            })
        }

    }
}


# Visualize artifacts over time -------------------------------------------


index <- seq(1, length(img_list), length.out = 16)
images <- img_list[index]

par(mfrow = c(4,4), mar = rep(0.2, 4))
rasterize <- function(x) {
    as.raster(x[1, , ])
}
images %>%
    purrr::map(rasterize) %>%
    purrr::iwalk(~{plot(.x)})


